### ISSUE TITLE
Overriding application.properties outside packaged jar in application.yml problem

### ISSUE THREAD

**Issue by nodje** [spring-projects/spring-boot#5844]
The problem occurs both for 1.3.3 and 1.4.0.M2

I'm using the `application.yml` format for app configuration.

While trying to put reusable default configuration in an external jar so that could be shared between apps, I stumble on the doc [boot-features-external-config-yaml-shortcomings](http://docs.spring.io/spring-boot/docs/current/reference/html/boot-features-external-config.html#boot-features-external-config-yaml-shortcomings) section.
So I switched to an `application.properties` format. So far so good, but when I need to override theses settings in an application using `application.yml` it ONLY works if the overriding settings are put in the profile section of the `application.yml`.

this works:
```
---
spring:
    profiles: dev
app:
    metrics:
        logs:
            enabled: true
```
this doesn't:
```
app:
    metrics:
        logs:
            enabled: true
---
spring:
    profiles: dev
```

hope this is clear. 
Basically the second section should work so that the default for all profiles could be locally override on a per profile basis.

**Comment by snicoll** [comment]
The shortcomings section has no link to what you're describing here as far as I can see. You can load `application.yml` from the same locations as `application.properties` and with the same feature. That section states that if you are using `@PropertySource` you can't use it with yaml files. That's not what you're doing, are you?

**Comment by nodje** [comment]
True, the section I indicated is not directly linked to what I'm describing.
But I've been experimenting loading an `application.yml` from an external jar without success, also without success trying with `@PropertySource("commons.yml")` so I guess I associated the two problem as being due to this limitation.
Anyway I switched back to the `.properties` format, which is working, but with the limitation I indicated.

**Comment by wilkinsona** [comment]
@nodje I'm not sure that I understand your description of the behaviour, or what the problem is. Can you please provide a small sample with some tests that illustrate the problem?

**Comment by nodje** [comment]
It'd be difficult for me to provide it in the coming days. I'll try though if needed.
To reiterate on the problem:
Structure:
```
common.jar
\application.properties
   app.metrics.log.enable=false
```
```
main-application.war -->dependency common.jar
\application.yml
```
I need to override the `app.metrics.log.enabled` in my main app's `application.yml`
this works:
```
---
spring:
    profiles: dev
app:
    metrics:
        logs:
            enabled: true
```
this doesn't:
```
app:
    metrics:
        logs:
            enabled: true
---
spring:
    profiles: dev
```
Only the profile section of the `application.yml` take into account the overriding property, not the global/default section

**Comment by wilkinsona** [comment]
It's still not clear to me exactly where the properties and YAML files are. I can't tell if they are inside or outside the jar in each case. Without an accurate reproduction, we'll just be guessing about the problem and we don't have the time to do that.

If you'd like us to spend some time investigating, please spend some time providing a sample that reproduces the problem.

**Comment by spring-issuemaster** [comment]
If you would like us to look at this issue, please provide the requested information. If the information is not provided within the next 7 days this issue will be closed.

**Comment by nodje** [comment]
Please have a look at the https://github.com/nodje/spring-boot-external-property-override-issue.git project, which reproduce this edge case bug.

**Comment by wilkinsona** [comment]
Thanks for the sample. It's made it much clearer what's going on.

This is working as designed. You have two configuration files that are both on the classpath: `application.properties` in `common` and `application.yml` in `prop-overriding`. As described in the [documentation on externalised configuration](http://docs.spring.io/spring-boot/docs/current/reference/htmlsingle/#boot-features-external-config) there are both "application properties packaged inside your jar (application.properties and YAML variants)". When loaded from the same location (the root of the classpath in this case), properties files take precedence over YAML files

If you want to override the properties that are configured in `application.properties` from `common` then you need to relocate your `application.yml` file. As [described in the documentation](http://docs.spring.io/spring-boot/docs/current/reference/htmlsingle/#boot-features-external-config-application-property-files) you have a few options. If you want to keep it packaged inside `prop-overriding` then move it into `src/main/resources/config`. Alternatively, you can move it outside of the project into `./config` or `./`.

**Comment by nodje** [comment]
Thanks for your answer Andy.
Though I guess I still don't understand how this can be the expected behavior.

If there was no way to override the `common` `application.properties` with an `application.yml` in the main application at the default location, then yeah, fair enough. You need to put the `application.yml` in `src/main/resources/config`, period.

But that's not the case: it actually works when using *profiled* properties from `application.yml` but not the root/default profile properties. This doesn't make sense to me. It should be all or nothing IMHO.

This is how I understand the [24. Externalized Configuration](http://docs.spring.io/spring-boot/docs/current/reference/htmlsingle/#boot-features-external-config) description of the `PropertySource` precedence.

**Comment by wilkinsona** [comment]
Profiled properties take precedence over normal properties. When you run with the dev profile, the profiled property in application.yml wins over the normal property in application.properties. This is 8 vs 10 in the list in the documentation. When you run without a profile, the properties in application.properties and application.yml that apply are both normal so it's 10 vs 10 in the list. As a tiebreaker the properties file wins.

**Comment by nodje** [comment]
OK, got it now.
I somehow always understood the `Application properties outside of your packaged jar` as being `outside of the main application jar` but still in a dependency jar file. 
I understand now that it means outside of a jar, as a file in the file system.

It'd be nice to have an order of precedence between `application.properties` in main jar/war and dependencies jars.

Thanks very much for your time.
